0.36.0 incorrectly detects posix_memalign on MSYS2 (native MINGW64 and MINGW32 targets/shells/environs)

Reporting on @ebassi's behalf, since they're busy. This is affecting downstream at ebassi/graphene#76.

Meson incorrectly detects the presence of `posix_memalign()` with the following in the project's `meson.build`:

```
conf.set('HAVE_POSIX_MEMALIGN', cc.has_function('posix_memalign', prefix: '#include <stdlib.h>'))
```


The output in `meson-logs/meson-log.txt` seems to indicate that it's settling on `__builtin_posix_memalign` compiling, and using that as a substitute for the real thing. But I have _never_ used meson before, so I don't really understand this.

```
[...]
Checking for function "memalign": NO
Running compile:
Working directory:  C:/Users/TESTUS~1/AppData/Local/Temp/tmp2yaurewq
Command line:  cc C:/Users/TESTUS~1/AppData/Local/Temp/tmp2yaurewq/testfile.c -std=gnu99 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -O0 -o C:/Users/TESTUS~1/AppData/Local/Temp/tmp2yaurewq/output.exe 

Code:
 #include <limits.h>
#include <stdlib.h>

        #if defined __stub_posix_memalign || defined __stub___posix_memalign
        fail fail fail this function is not going to work
        #endif
        
int main() { void *a = (void*) &posix_memalign; }
Compiler stdout:
 C:/Users/TESTUS~1/AppData/Local/Temp/tmp2yaurewq/testfile.c: In function 'main':

C:/Users/TESTUS~1/AppData/Local/Temp/tmp2yaurewq/testfile.c:8:33: error: 'posix_memalign' undeclared (first use in this function)

 int main() { void *a = (void*) &posix_memalign; }

                                 ^~~~~~~~~~~~~~

C:/Users/TESTUS~1/AppData/Local/Temp/tmp2yaurewq/testfile.c:8:33: note: each undeclared identifier is reported only once for each function it appears in


Compiler stderr:
 
Running compile:
Working directory:  C:/Users/TESTUS~1/AppData/Local/Temp/tmpk7achqy2
Command line:  cc C:/Users/TESTUS~1/AppData/Local/Temp/tmpk7achqy2/testfile.c -std=gnu99 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -O0 -o C:/Users/TESTUS~1/AppData/Local/Temp/tmpk7achqy2/output.exe 

Code:
 int main() { __builtin_posix_memalign; }
Compiler stdout:
 
Compiler stderr:
 
Checking for function "posix_memalign": YES
[...]
```

The function is not defined in any header the MINGW64 and MINGW32 builds should be using.

```
Test User@win7test MINGW64 /usr/src/graphene
$ grep -R posix_memalign /mingw{64,32}/include /usr/include

Test User@win7test MINGW64 /usr/src/graphene
$ find /mingw{64,32}/include /usr/include -name 'stdlib.*'
/mingw64/include/c++/6.2.0/stdlib.h
/mingw64/include/c++/6.2.0/tr1/stdlib.h
/mingw32/include/c++/6.2.0/stdlib.h
/mingw32/include/c++/6.2.0/tr1/stdlib.h
```

Confirmed not a dup. I'm completely lost now, other than to confirm the version I'm using:

```
Test User@win7test MINGW64 /usr/src/graphene/_build
$ gcc --version
gcc.exe (Rev2, Built by MSYS2 project) 6.2.0
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Test User@win7test MINGW64 /usr/src/graphene/_build
$ pacman -Qo `which gcc.exe`
/mingw64/bin/gcc.exe is owned by mingw-w64-x86_64-gcc 6.2.0-2

```

And the fact that #1066 does not fix it like @nirbheek said:

```
Compiler stdout:
 C:/Users/TESTUS~1/AppData/Local/Temp/tmp29u5wr84/testfile.c: In function 'main'
:
C:/Users/TESTUS~1/AppData/Local/Temp/tmp29u5wr84/testfile.c:8:33: error: 'posix_
memalign' undeclared (first use in this function)
 int main() { void *a = (void*) &posix_memalign; long b = (long) a; return (int)
 b; }
                                 ^~~~~~~~~~~~~~
C:/Users/TESTUS~1/AppData/Local/Temp/tmp29u5wr84/testfile.c:8:33: note: each und
eclared identifier is reported only once for each function it appears in
C:/Users/TESTUS~1/AppData/Local/Temp/tmp29u5wr84/testfile.c:8:58: warning: cast
from pointer to integer of different size [-Wpointer-to-int-cast]
 int main() { void *a = (void*) &posix_memalign; long b = (long) a; return (int)
 b; }
                                                          ^

Compiler stderr:

Running compile:
Working directory:  C:/Users/TESTUS~1/AppData/Local/Temp/tmp8zkcy2jf
Command line:  cc C:/Users/TESTUS~1/AppData/Local/Temp/tmp8zkcy2jf/testfile.c -std=gnu99 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -O0 -o C:/Users/TESTUS~1/AppData/Local/Temp/tmp8zkcy2jf/output.exe

Code:
 int main() { __builtin_posix_memalign; }
Compiler stdout:

Compiler stderr:

Checking for function "posix_memalign": YES
```

--- a/mesonbuild/compilers.py
+++ b/mesonbuild/compilers.py
@@ -1014,11 +1014,23 @@ int main(int argc, char **argv) {
         if self.links(templ.format(prefix, funcname), env, extra_args, dependencies):
             return True
         # Some functions like alloca() are defined as compiler built-ins which
-        # are inlined by the compiler, so test for that instead. Built-ins are
-        # special functions that ignore all includes and defines, so we just
-        # directly try to link via main().
-        return self.links('int main() {{ {0}; }}'.format('__builtin_' + funcname),
-                          env, extra_args, dependencies)
+        # are inlined by the compiler, so look for __builtin_symbol in the libc
+        # if there's no #include-s in prefix which would've #define-d the
+        # symbol correctly. If there is a #include, just check for the symbol
+        # directly. This is needed because the above #undef fancy footwork
+        # doesn't work for builtins.
+        # This fixes instances such as #1083 where MSYS2 defines
+        # __builtin_posix_memalign in the C library but doesn't define
+        # posix_memalign in the headers to point to that builtin which results
+        # in an invalid detection.
+        if '#include' not in prefix:
+            code = 'int main() {{ {0}; }}'
+            return self.links(code.format('__builtin_' + funcname), env,
+                              extra_args, dependencies)
+        else:
+            code = '{0}\n' + stubs_fail + '\nint main() {{ {1}; }}'
+            return self.links(code.format(prefix, funcname), env, extra_args,
+                              dependencies)
 
     def has_members(self, typename, membernames, prefix, env, extra_args=None, dependencies=None):
         if extra_args is None:

--- a/test cases/common/43 has function/meson.build
+++ b/test cases/common/43 has function/meson.build
@@ -1,5 +1,7 @@
 project('has function', 'c', 'cpp')
 
+host_system = host_machine.system()
+
 # This is used in the `test_compiler_check_flags_order` unit test
 unit_test_args = '-I/tmp'
 compilers = [meson.get_compiler('c'), meson.get_compiler('cpp')]
@@ -34,7 +36,7 @@ foreach cc : compilers
   # We can't check for the C library used here of course, but if it's not
   # implemented in glibc it's probably not implemented in any other 'slimmer'
   # C library variants either, so the check should be safe either way hopefully.
-  if host_machine.system() == 'linux' and cc.get_id() == 'gcc'
+  if host_system == 'linux' and cc.get_id() == 'gcc'
     assert (cc.has_function('poll', prefix : '#include <poll.h>',
                             args : unit_test_args),
             'couldn\'t detect "poll" when defined by a header')
@@ -42,6 +44,12 @@ foreach cc : compilers
     assert (not cc.has_function('lchmod', prefix : lchmod_prefix,
                                 args : unit_test_args),
             '"lchmod" check should have failed')
+    # Check that built-ins are found properly both with and without headers
+    assert(cc.has_function('alloca', args : unit_test_args),
+           'built-in alloca must be found on Linux')
+    assert(cc.has_function('alloca', prefix : '#include <alloca.h>',
+           args : unit_test_args),
+           'built-in alloca must be found on Linux with #include')
   endif
 
   # For some functions one needs to define _GNU_SOURCE before including the

